File: backend/models/overlay.py
from typing import Literal

from pydantic import BaseModel, Field

OverlayMode = Literal["assistant", "translation"]
OverlayPhase = Literal["hidden", "idle", "loading", "ready", "error"]


class OverlayReadingContext(BaseModel):
    resource_url: str = Field(default="", max_length=4096)
    resource_title: str = Field(default="", max_length=1024)
    section_heading: str = Field(default="", max_length=1024)
    context_before: str = Field(default="", max_length=4000)
    context_after: str = Field(default="", max_length=4000)
    source_kind: str = Field(default="browser_selection", max_length=128)


class OverlayStateResponse(OverlayReadingContext):
    revision: int
    visible: bool
    mode: OverlayMode = "assistant"
    phase: OverlayPhase
    context_id: str = ""
    source_text: str = ""
    translated_text: str = ""
    source_language: str = "auto"
    target_language: str = "zh-CN"
    provider: str = ""
    message: str = ""
    translation_notice: str = ""
    companion_conversation_id: str = ""


class OverlayAssistantRequest(OverlayReadingContext):
    context_id: str = Field(min_length=1, max_length=128)
    source_text: str = Field(min_length=1, max_length=20_000)
    source_language: str = "auto"
    target_language: str = "zh-CN"


class OverlayLoadingRequest(OverlayReadingContext):
    context_id: str = Field(min_length=1, max_length=128)
    source_text: str = Field(min_length=1, max_length=20_000)
    source_language: str = "auto"
    target_language: str = "zh-CN"


class OverlayPresentRequest(OverlayLoadingRequest):
    translated_text: str = Field(min_length=1, max_length=50_000)
    provider: str = "unknown"
    translation_notice: str = Field(default="", max_length=500)


class OverlayErrorRequest(OverlayReadingContext):
    context_id: str = Field(min_length=1, max_length=128)
    source_text: str = Field(default="", max_length=20_000)
    source_language: str = "auto"
    target_language: str = "zh-CN"
    message: str = Field(default="Translation failed", max_length=500)


class OverlayCompanionBindingRequest(BaseModel):
    context_id: str = Field(min_length=1, max_length=128)
    conversation_id: str = Field(default="", max_length=128)

File: apps/desktop/src/api/overlay.ts
import { apiGet, apiPost } from "./client"
import { getCachedReadingSelection } from "./reading"
import type {
  OverlayErrorRequest,
  OverlayLoadingRequest,
  OverlayPresentRequest,
  OverlayStateResponse,
  ReadingContextFields,
} from "./types"
import { desktop } from "../desktop"

export type OverlayMode = "assistant" | "translation"
export type UnifiedOverlayStateResponse = OverlayStateResponse & {
  mode: OverlayMode
  translation_notice: string
  companion_conversation_id?: string
}

export interface OverlayAssistantRequest extends Partial<ReadingContextFields> {
  context_id: string
  source_text: string
  source_language: string
  target_language: string
}

export interface UnifiedOverlayPresentRequest extends OverlayPresentRequest {
  translation_notice?: string
}

const emptyReadingContext: ReadingContextFields = {
  resource_url: "",
  resource_title: "",
  section_heading: "",
  context_before: "",
  context_after: "",
  source_kind: "",
}

function readingContextFor(contextId: string): ReadingContextFields {
  const selection = getCachedReadingSelection()
  if (!selection || selection.selection_id !== contextId) return emptyReadingContext

  return {
    resource_url: selection.resource_url,
    resource_title: selection.resource_title,
    section_heading: selection.section_heading,
    context_before: selection.context_before,
    context_after: selection.context_after,
    source_kind: selection.source_kind,
  }
}

function withReadingContext<T extends { context_id: string }>(payload: T): T & ReadingContextFields {
  return {
    ...readingContextFor(payload.context_id),
    ...payload,
  }
}

async function notifyOverlayStateChanged(contextId = ""): Promise<void> {
  try {
    await desktop.overlay.notifyStateChanged(contextId)
  } catch {
    // Event delivery is an optimization; polling remains the recovery path.
  }
}

export function getOverlayState(): Promise<UnifiedOverlayStateResponse> {
  return apiGet<UnifiedOverlayStateResponse>("/api/overlay")
}

export async function showOverlayAssistant(payload: OverlayAssistantRequest): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<UnifiedOverlayStateResponse, OverlayAssistantRequest & ReadingContextFields>(
    "/api/overlay/assistant",
    withReadingContext(payload),
  )
  await notifyOverlayStateChanged(payload.context_id)
  return response
}

export async function showOverlayLoading(payload: OverlayLoadingRequest): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<UnifiedOverlayStateResponse, OverlayLoadingRequest & ReadingContextFields>(
    "/api/overlay/loading",
    withReadingContext(payload),
  )
  await notifyOverlayStateChanged(payload.context_id)
  return response
}

export async function presentOverlay(payload: UnifiedOverlayPresentRequest): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<UnifiedOverlayStateResponse, UnifiedOverlayPresentRequest & ReadingContextFields>(
    "/api/overlay/present",
    withReadingContext(payload),
  )
  await notifyOverlayStateChanged(payload.context_id)
  return response
}

export async function showOverlayError(payload: OverlayErrorRequest): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<UnifiedOverlayStateResponse, OverlayErrorRequest & ReadingContextFields>(
    "/api/overlay/error",
    withReadingContext(payload),
  )
  await notifyOverlayStateChanged(payload.context_id)
  return response
}

export async function bindOverlayCompanionConversation(
  contextId: string,
  conversationId: string,
): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<
    UnifiedOverlayStateResponse,
    { context_id: string; conversation_id: string }
  >(
    "/api/overlay/companion",
    { context_id: contextId, conversation_id: conversationId },
  )
  await notifyOverlayStateChanged(contextId)
  return response
}

export async function dismissOverlay(): Promise<UnifiedOverlayStateResponse> {
  const response = await apiPost<UnifiedOverlayStateResponse, Record<string, never>>("/api/overlay/dismiss", {})
  await notifyOverlayStateChanged("")
  return response
}
